http.py

```diff
@@ -2,138 +2,138 @@
-
-    def clear(self):
-        self.parser = HttpParser()
-
-    def parse(self, data):
-        self.parser.execute(data, len(data))
-
-        if self.parser.is_message_complete():
-            self.version = "{0}.{1}".format(self.parser.get_version()[0], self.parser.get_version()[1])
-            self.status = int(self.parser.get_status_code())
-            self.method = self.parser.get_method()
-            self.url = self.parser.get_url()
-            self.path = self.parser.get_path()
-            self.query_string = self.parser.get_query_string()
-            self.header = self.parser.get_headers()
-            self.body = self.parser.recv_body()
-
-    def serialize(self):
-        data = {}
-        data["version"] = self.version
-        data["status"] = self.status
-        data["method"] = self.method
-        data["url"] = self.url
-        data["path"] = self.path
-        data["query_string"] = self.query_string
-        data["header"] = self.header
-        data["body"] = self.body.encode("base64")
-        return data
-
-    def deserialize(self, data):
-        self.version = data["version"]
-        self.status = data["status"]
-        self.method = data["method"]
-        self.url = data["url"]
-        self.path = data["path"]
-        self.query_string = data["query_string"]
-        self.header = data["header"]
-        self.body = data["body"].decode("base64")
-
-class HttpRequest(HttpMessage):
+        self.version = version 
+        self.status = status
+        self.method = method
+        self.url = url
+        self.path = path
+        self.query_string = query_string
+        self.header = header
+        self.body = body
+
+class HttpMessageBuilder(object):
     def __init__(self):
-        super(HttpRequest, self).__init__()
-    
-    def _assemble_header(self):
-        http_header_query_str = "{0} {1} HTTP/{2}".format(self.method, self.url, self.version)
-        http_header_fields = [ "{0}: {1}".format(key, self.header[key]) for key in self.header ]
-
-        http_header_list = [http_header_query_str]
-        http_header_list.extend(http_header_fields)
-
-        return "{0}\r\n\r\n".format("\r\n".join(http_header_list))
+        super(HttpMessageBuilder, self).__init__()
+        self.http_parser = HttpParser()
 
-    def _assemble_body(self):
-        raise NotImplementedError
-
-    def _assemble_data(self):
-        raise NotImplementedError
+    def parse(self, data):
+        self.http_parser.execute(data, len(data))
 
     @property
-    def data(self):
-        final_data = b"{0}{1}".format(self._assemble_header(), self.body)
-        return final_data
-
-class HttpResponse(HttpMessage):
-    def __init__(self):
-        super(HttpResponse, self).__init__()
-
-    def is_chunked_encoding(self):
+    def is_done(self):
+        return self.http_parser.is_message_complete()
+
+    def build(self):
+        if not self.is_done:
+            raise IOError
```

Comment:
This IOError maybe need change to AssertionError.
We assert that when calling the method build, need call is_done to check is done or not first.
But if not call, we should fail fast with an error/exception, that the method had been used in a wrong way.